docs/src/content/en/docs/agents/using-tools-and-mcp.mdx

```diff
@@ -107,6 +112,41 @@ const agent = new Agent({
 });
 ```
 
+When creating agents that will consume an MCP server in the same repo they need to connect to, always use promise-based (lazy) tools to prevent race conditions.
+
+```typescript filename="src/mastra/agents/selfReferencingAgent.ts"
+import { Agent } from "@mastra/core/agent";
+import { MCPServer } from "@mastra/mcp";
+import { MCPClient } from "@mastra/mcp";
+import { openai } from "@ai-sdk/openai";
+
+const myAgent = new Agent({
+  name: "My Agent",
+  description: "An agent that can use tools from an http MCP server", 
+  instructions: "You can use remote calculation tools.",
+  model: openai("gpt-4o-mini"),
+  tools: async () => {
+    // Tools resolve lazily when needed, not during initialization
+    const mcpClient = new MCPClient({
+      servers: {
+        myServer: {
+          url: new URL("http://localhost:3000/mcp"),
+        },
+      },
+    });
+    return await mcpClient.getTools();
+  },
```

Comment:
I was hoping we'd be able to pass an unresolved promise directly like this:
```suggestion
  tools: mcpClient.getTools()
```

Does that work with this change?

Reply:
The main difference is with a direct promise you only call `getTools` one time, and it'll likely start up a lot quicker